core/src/main/java/org/springframework/security/access/method/MethodInvocationAuthorizationManagerResolver.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.access.method;
+
+import java.lang.reflect.Method;
+
+import org.aopalliance.intercept.MethodInvocation;
+
+import org.springframework.security.authorization.AuthorizationManager;
+
+/**
+ * A resolver that is used for obtaining an {@link AuthorizationManager} instance for the
+ * check before the method invocation.
+ *
+ * @author Evgeniy Cheban
+ */
+public interface MethodInvocationAuthorizationManagerResolver {
```

Comment:
It feels odd to have an interface that returns an `AuthorizationManager` that could possibly mutate the method parameters. The JavaDoc for `AuthorizationManager` says:

```
Determines if access should be granted for a specific authentication and object.
```

Instead of `AuthorizationManagerResolver`, `AfterAuthorizationManagerResolver`, etc., you might consider the advice abstraction:

```java
public interface AuthorizationManagerBeforeAdvice<T> extends AuthorizationManager<T>, Pointcut {
}

public interface AuthorizationManagerAfterAdvice<T> extends Pointcut {
    Object check(Supplier<Authentication> authentication, T object, Object returnedObject);
}
```

The `T` might be `MethodAuthorizationContext` -- similar in nature to `RequestAuthorizationContext` -- that contains the `MethodInvocation` and the target class derived from `AopUtils`.

These interfaces are nice because they remove one layer of indirection (the resolver) and they communicate that implementations are at liberty to mutate the method parameters and return values.

Also, if the advice extends `Pointcut`, it clearly communicates that implementers need to describe when their advice applies. In that case, the pointcut advisor can consult the advice directly instead of attempting to resolve an authorization manager. This also allows implementations to use AOP support like `AnnotationMethodMatcher`. It would be interesting to see if there's some appetite for enhancing AOP's `MethodMatcher` in the same way that Security's `RequestMatcher` can return a `MatchResult`.

This more closely aligns with `PrePostAdviceReactiveMethodInterceptor` which will likely make for an easier transition in #9401.

Reply:
One other application to think about is https://github.com/spring-projects/spring-hateoas/issues/1343. It would be nice if it were easy for Spring HATEOAS to make a mock authorization decision, perhaps like this:

```java
@Autowired
AuthorizationManager<MethodInvocation> authorizationManager;

// ...
SimpleMethodInvocation methodInvocation = ...
if (authorizationManager.check(methodInvocation).isGranted()) {
    // ... include HATOEAS link
}
```

It's nice that Spring Security publishing a `AuthorizationManagerBeforeAdvice<MethodInvocation>` `@Bean` can achieve that.

Reply:
I updated the PR with this solution.